packages/api-client-core/src/InternalModelManager.ts

```diff
@@ -194,16 +194,16 @@ export class InternalModelManager {
     this.capitalizedApiIdentifier = camelize(apiIdentifier);
   }
 
-  async findOne(id: string, throwOnEmptyData = true): Promise<GadgetRecord<any>> {
+  async findOne(id: string, throwOnEmptyData = true): Promise<GadgetRecord<RecordShape>> {
     const response = await this.connection.currentClient.query(internalFindOneQuery(this.apiIdentifier), { id }).toPromise();
     const assertSuccess = throwOnEmptyData ? assertOperationSuccess : assertNullableOperationSuccess;
     const result = assertSuccess(response, ["internal", this.apiIdentifier]);
-    return hydrateRecord(response, result);
+    return hydrateRecord<RecordShape>(response, result);
   }
 
-  async maybeFindOne(id: string): Promise<GadgetRecord<any> | null> {
+  async maybeFindOne(id: string): Promise<GadgetRecord<RecordShape> | null> {
     const record = await this.findOne(id, false);
-    return record ?? null;
+    return "id" in record ? record : null;
```

Comment:
So not as is, BUT... now that I think about it, `findOne` is going to throw on empty anyway, right (rhetorical question — yes, it will)? So if I change the logic there to skip hydration on null (and thus change the return to include null as an option), then yes, I can do what you're saying.

I didn't want to touch the hydration part at first, but it shouldn't affect the behaviour at all, and it will be more robust to changes, as you say. I'll do that then.

Reply:
If you're up for it, you can also tighten up the return types by overloading the function signature for when throwOnEmptyData is true/false, but I'm game to come back to that later 🚀